packages/stack/proxy/src/rpc-manager.ts

```diff
@@ -0,0 +1,196 @@
+      this._web3 = null;
+
+      // web.eth.getAccounts may return a different value now
+      // update accounts across all modifiers
+      this.resetAccounts(cb);
+    });
+
+    this.embark.registerActionForEvent("blockchain:proxy:request", this.onProxyRequest.bind(this));
+    this.embark.registerActionForEvent("blockchain:proxy:response", this.onProxyResponse.bind(this));
+  }
+
+  private setCommandHandlers() {
+    this.events.setCommandHandler("rpc:accounts:reset", this.resetAccounts.bind(this));
+
+    this.events.setCommandHandler("rpc:request:interceptor:register", this.registerRequestInterceptor.bind(this));
+    this.events.setCommandHandler("rpc:response:interceptor:register", this.registerResponseInterceptor.bind(this));
+  }
+
+  private async onProxyRequest<T>(params: ProxyRequestParams<T>, callback: Callback<ProxyRequestParams<T>>) {
+    try {
+      params = await this.executeInterceptors(this.requestInterceptors, params);
+    } catch (err) {
+      err.message = __("Error executing RPC request modifications for '%s': %s", params?.request?.method, err.message);
+      return callback(err);
+    }
+    callback(null, params);
+  }
+
+  private async onProxyResponse<T, R>(params: ProxyResponseParams<T, R>, callback: Callback<ProxyResponseParams<T, R>>) {
+    try {
+      params = await this.executeInterceptors(this.responseInterceptors, params);
+    } catch (err) {
+      err.message = __("Error executing RPC response modifications for '%s': %s", params?.request?.method, err.message);
+      return callback(err);
+    }
+    callback(null, params);
+  }
+
+  private registerRequestInterceptor<T>(filter: string | RegExp, interceptor: RpcRequestInterceptor<T>, options: RegistrationOptions = { priority: 50 }, callback?: Callback<null>) {
+    this.requestInterceptors.push({ filter, options, interceptor });
+    if (callback) {
+      callback();
+    }
+  }
+
+  private registerResponseInterceptor<T, R>(filter: string | RegExp, interceptor: RpcResponseInterceptor<T, R>, options: RegistrationOptions = { priority: 50 }, callback?: Callback<null>) {
+    this.responseInterceptors.push({ filter, options, interceptor });
+    if (callback) {
+      callback();
+    }
+  }
+
+  private async resetAccounts(cb: Callback<null>) {
+    this._nodeAccounts = null;
+    this._accounts = null;
+    cb();
+  }
+
+  private async executeInterceptors<T>(
+    registrations: Array<RpcRequestRegistration<T>>,
+    params: ProxyRequestParams<T>
+  ): Promise<ProxyRequestParams<T>>;
+  private async executeInterceptors<T, R>(
+    registrations: Array<RpcResponseRegistration<T, R>>,
+    params: ProxyResponseParams<T, R>
+  ): Promise<ProxyResponseParams<T, R>>;
+  private async executeInterceptors<T, R>(
+    registrations: Array<RpcRequestRegistration<T> | RpcResponseRegistration<T, R>>,
+    params: ProxyRequestParams<T> | ProxyResponseParams<T, R>
+  ): Promise<ProxyRequestParams<T> | ProxyResponseParams<T, R>> {
```

Comment:
This might look overwhelming, but since we're using `executeInterceptors()` for both, `requestInterceptors` and `responseInterceptors`, we have to overload the method's signature here.

Reply:
Well done.